beets/metadata_plugins.py

```diff
@@ -386,6 +396,34 @@ def item_candidates(
             self.tracks_for_ids([result["id"] for result in results if result]),
         )
 
+    def _construct_search_query(
+        self, filters: SearchFilter, keywords: str = ""
+    ) -> str:
+        """Construct a query string with the specified filters and keywords to
+        be provided to the Spotify (or similar) Search API.
+
+        At the moment, this is used to construct a query string for:
+        - Spotify (https://developer.spotify.com/documentation/web-api/reference/search).
+        - Deezer (https://developers.deezer.com/api/search).
+
+        :param filters: Field filters to apply.
+        :param keywords: Query keywords to use.
+        :return: Query string to be provided to the Search API.
+        """
```

Comment:
The interface sets up this method, but it shouldn't know the specifics about what implements it. 

Since the parameters' names are clear, I don't think we need to repeat this information in the docstring.

```suggestion
        """Construct a query with the specified filters and query string."""
```

Reply:
I agree, the issue is tho that we lose the references to the docs if we just remove it, as these methods are present in the plugins anymore. There is no repeated information.

Reply:
This docstring sort of violates the [Open/Closed SOLID principle](https://en.wikipedia.org/wiki/SOLID#Open–closed_principle): if we add another data source that inherits `SearchApiMetadataSourcePlugin` and uses this method, we will need to update this docstring, which is bad practice. Practically speaking, this docstring would not get updated and end up being misleading.

This method (including the docstring) should be implemented/documented in a generic manner and closed for modification in order to prevent this issue.

You can move the URLs under respective `_search_api` implementations instead.

Reply:
Yeah, I can move the comment, but I still think it fits better alongside the query function. The root issue here is that this setup already violates the dependency inversion principle, so moving the comment doesn’t really solve anything, it just reinforces the design flaw.

> This docstring sort of violates the [Open/Closed SOLID principle](https://en.wikipedia.org/wiki/SOLID#Open%E2%80%93closed_principle): if we add another data source that inherits SearchApiMetadataSourcePlugin and uses this method, we will need to update this docstring, which is bad practice. Practically speaking, this docstring would not get updated and end up being misleading.

I get the concern about the docstring violating the Open/Closed Principle, and that’s valid, but we’re already operating in a structure that makes adhering to these principles tricky. Rather than selectively applying them, I think we should either address the underlying design or be explicit about the compromises we're making.

Reply:
I mainly think it is important for someone looking at the function to know why the query string looks that way. And this is only clear in the context of the spotify api documentation.

Reply:
I don't think I'm following you regarding these two points:

> The root issue here is that this setup already violates the dependency inversion principle 

> And this is only clear in the context of the spotify api documentation

Reply:
1. I've been trying to point out for quite some time that there's a problem with how plugins and core logic are not properly separated (see, for example, https://github.com/beetbox/beets/pull/5882#discussion_r2217014242 or the get_artists discussion). Our ongoing debates stem from this fundamental issue. Since you value design principles, this essentially comes down to a violation of the dependency inversion principle.

2. Consider encountering the `create_query` function somewhere without any context. A comment explaining why the query is constructed that way, such as linking to spotifys or deezers specifications, would be very helpful. Having to dive into the spotify or deezer plugin just to understand the rationale makes maintenance harder, even if adding such comments slightly bends some design principles.

Reply:
> 1. I've been trying to point out for quite some time that there's a problem with how plugins and core logic are not properly separated (see, for example,

I'm completely lost - I'm struggling to see what this means

>> plugin = beetsplug(beets)
>> \# vs
>> plugin = beets(beetsplug(beets))

nor how any of this relates to the question that this thread was started for: _should `_construct_search_query` method docstring in the base class provide links to Spotify and Deezer APIs?

> A comment explaining why the query is constructed that way

I completely agree with this!



> Having to dive into the spotify or deezer plugin just to understand the rationale makes maintenance harder

Agree, the docstring should contain all relevant information - this is exactly why I'm not expecting to see any external URLs there.

> such as linking to spotifys or deezers specifications, would be very helpful.

Following your idea above, we should not expect devs to navigate these URLs in order to understand what this method does.